File: 3152-special-array-ii/3152-special-array-ii.java
class Solution {
    public boolean[] isArraySpecial(int[] nums, int[][] q) {
        int[] spc=new int[nums.length];
        spc[0] = 0;
        int pp = nums[0]%2;
        int sm = spc[0];
        for(int i=1;i<nums.length;i++){
            int cp = nums[i]%2;
            if(cp==pp){
                spc[i] = i;
                sm = i;
            }
            else{
                spc[i] = sm;
            }
            pp=cp;
        }
        boolean[] ans=new boolean[q.length];
        for(int i=0;i<q.length;i++){
            if(spc[q[i][0]]==spc[q[i][1]]) ans[i]=true;
            else ans[i] = false;
        }
        return ans;
    }
}

File: 3152-special-array-ii/3152-special-array-ii.py
class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        n = len(nums)
        ps = [0] * n

        for i in range(1, n):
            ps[i] = nums[i] % 2 == nums[i - 1] % 2
            if i:
                ps[i] += ps[i - 1]
        
        ans = []
        for v in queries:
            l, r = v
            sum_val = ps[r] - ps[l]
            ans.append(sum_val == 0)

        return ans